CacheTools.php

```diff
@@ -1,163 +1,169 @@
+* If you did not receive a copy of the license and are unable to
+* obtain it through the world-wide-web, please send an email
+* to license@prestashop.com so we can send you a copy immediately.
+*
+* DISCLAIMER
+*
+* Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+* versions in the future. If you wish to customize PrestaShop for your
+* needs please refer to http://www.prestashop.com for more information.
+*
+*  @author PrestaShop SA <contact@prestashop.com>
+*  @copyright  2007-2011 PrestaShop SA
+*  @license    http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
+*  International Registered Trademark & Property of PrestaShop SA
+*/
+
+class CacheTools
+{
+	public static function checkProductCache($ids_product, $region, Cart $cart)
+	{
+		$result = Db::getInstance()->ExecuteS('SELECT ac.`tax_rate`, ac.`update_date`
+		FROM `'._DB_PREFIX_.'avalara_product_cache` ac
+		WHERE ac.`id_product` IN ('.pSQL($ids_product).') AND ac.`region` = \''.pSQL($region).'\' AND ac.`id_address` = '.(int)$cart->{Configuration::get('PS_TAX_ADDRESS_TYPE')});
+
+		if (count($result) == count($cart->getProducts()))
+		{
+			// Compare date/time
+			date_default_timezone_set(@date_default_timezone_get());
+			$date2 = time();
+			foreach ($result as $line)
+				if (abs($date2 - strtotime($line['update_date'])) > Configuration::get('AVALARA_CACHE_MAX_LIMIT'))
+					return true;
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * @brief Check the Carrier Cache
+	 *
+	 * @param cart Cart object
+	 *
+	 * @return True if the cache expired, false otherwise
+	 */
+	public static function checkCarrierCache(Cart $cart)
+	{
+		$tmp = array('id_carrier' => (int)$cart->id_carrier, 'id_address' => (int)$cart->{Configuration::get('PS_TAX_ADDRESS_TYPE')});
+		foreach ($cart->getProducts() as $product)
+			$tmp[] = array('id_product' => (int)$product['id_product'], 'id_product_attribute' => (int)$product['id_product_attribute'], 'quantity' => (int)$product['quantity']);
+			
+		$result = Db::getInstance()->getRow('
+		SELECT `cart_hash`, `update_date`
+		FROM `'._DB_PREFIX_.'avalara_carrier_cache`
+		WHERE `cart_hash` = \''.pSQL(md5(serialize($tmp))).'\'
+		ORDER BY update_date DESC');
+
+		if (!$result)
+			return true;
+
+		// Compare date/time
+		date_default_timezone_set(@date_default_timezone_get());
+		if (abs(time() - strtotime($result['update_date'])) > Configuration::get('AVALARA_CACHE_MAX_LIMIT'))
+			return true;
+
+		return false;
+	}
+
+	public static function updateProductsTax(AvalaraTax $avalaraModule, Cart $cart, $id_address, $region, $taxable = true)
+	{
+		$p = array();
+		foreach ($cart->getProducts() as $product)
+		{
+			$avalaraProducts = array('id_product' => (int)$product['id_product'],
+												 'name' => $product['name'],
+												 'description_short' => $product['description_short'],
+												 'quantity' => 1, // This is a per product, so qty is 1
+												 'total' => (float)1000000,
```

Comment:
This is because getting the tax from avalara asks for a total dollar amount tax instead of a rate. If the shopper would perhaps try to purchase 10,000 one-cent products, the total should be taxed based on the total amount, not a per product tax as any tax below 50% would round down to zero. This issue could cause rounding errors for any number of products (usually in the order of a few cents as most products are several dollars or more) until the price approaches $10,000, however I chose an even higher number in my test to be doubly sure that rounding errors do not occur. A fundamental change in how it is requested and cached is needed since the rate is more useful than a total cost, but since the API only provides a total dollar amount tax, this is the best workaround.